src/What4/ExprHelpers.hs

```diff
@@ -402,3 +428,641 @@ mkSafeAsserts ::
 mkSafeAsserts sym (W4G.GroundEvalFn fn) = do
   cache <- W4B.newIdxCache
   return $ W4G.GroundEvalFn (\e -> stripAsserts sym cache e >>= fn)
+
+newtype SetF f tp = SetF (Set (AsOrd f tp))
+
+type ExprSet sym = SetF (W4.SymExpr sym)
+
+data AsOrd f tp where
+  AsOrd :: OrdF f => { unAsOrd :: f tp } -> AsOrd f tp
+
+instance Eq (AsOrd f tp) where
+  (AsOrd a) == (AsOrd b) = case compareF a b of
+    EQF -> True
+    _ -> False
+
+instance Ord (AsOrd f tp) where
+  compare (AsOrd a) (AsOrd b) = toOrdering $ compareF a b
+
+deriving instance Semigroup (SetF sym tp)
+deriving instance Monoid (SetF sym tp)
+
+
+singletonExpr :: OrdF (W4.SymExpr sym) => W4.SymExpr sym tp -> ExprSet sym tp
+singletonExpr e = SetF (S.singleton (AsOrd e))
+
+listToExprSet :: W4.IsSymExprBuilder sym => [W4.SymExpr sym tp] -> ExprSet sym tp
+listToExprSet l = SetF $ S.fromList $ map AsOrd l
+
+exprSetToList :: ExprSet sym tp -> [W4.SymExpr sym tp]
+exprSetToList (SetF es) = map unAsOrd $ S.toList es
+
+exprSetSize :: ExprSet sym tp -> Int
+exprSetSize (SetF es) = S.size es
+
+inExprSet :: OrdF (W4.SymExpr sym) => W4.SymExpr sym tp -> ExprSet sym tp -> Bool
+inExprSet e (SetF es) = S.member (AsOrd e) es
+
+
+-- | Compute a predicate representing the path condition of the
+-- expression according to its internal mux structure.
+-- i.e. given the model @[x := 2, y := 5]@ and the expression
+-- @if x > y then (x - y) else (y - x)@,
+-- it will produce the predicate @x < y@.
+-- Recurses into the sub-structure of predicates to yield a minimal result.
+-- i.e. given @if A OR B OR C then D else E@ and a model that sets @A@ to true,
+-- but @B@ and @C@ as false, the path condition will contain only @A@.
```

Comment:
The specific values don't matter, only the path through the mux structure that they represent. So in this case any `x` and `y` would yield the same path condition as long as `x < y`, since we're just looking at the predicate.